components/constellation/constellation.rs

```diff
@@ -219,30 +219,47 @@ pub struct InitialConstellationState {
     pub webrender_api_sender: Option<webrender_traits::RenderApiSender>,
 }
 
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct FrameState {
+    instant: Instant,
+    pipeline_id: PipelineId,
+    frame_id: FrameId,
+}
+
+impl FrameState {
+    fn new(pipeline_id: PipelineId, frame_id: FrameId) -> FrameState {
+        FrameState {
+            instant: Instant::now(),
+            pipeline_id: pipeline_id,
+            frame_id: frame_id,
+        }
+    }
+}
+
 /// Stores the navigation context for a single frame in the frame tree.
 struct Frame {
     id: FrameId,
-    prev: Vec<(PipelineId, Instant)>,
-    current: (PipelineId, Instant),
-    next: Vec<(PipelineId, Instant)>,
+    prev: Vec<FrameState>,
+    current: FrameState,
```

Comment:
hmm... what's the point of having `frame.id` now? Isn't it equivalent to `frame.current.id`?

Maybe worth trimming that in a followup.